src/pounder/dds_output.rs

```diff
@@ -0,0 +1,107 @@
+///! The DdsOutput is used as an output stream to the pounder DDS.
+use super::QspiInterface;
+use crate::hrtimer::HighResTimerE;
+use ad9959::{Channel, DdsConfig, ProfileSerializer};
+use stm32h7xx_hal as hal;
+
+/// The DDS profile update stream.
+pub struct DdsOutput {
+    _qspi: QspiInterface,
+    io_update_trigger: HighResTimerE,
+    config: DdsConfig,
+}
+
+impl DdsOutput {
+    /// Construct a new DDS output stream.
+    ///
+    /// # Note
+    /// It is assumed that the QSPI stream and the IO_Update trigger timer have been configured in a
+    /// way such that the profile has sufficient time to be written before the IO_Update signal is
+    /// generated.
+    ///
+    /// # Args
+    /// * `qspi` - The QSPI interface to the run the stream on.
+    /// * `io_update_trigger` - The HighResTimerE used to generate IO_Update pulses.
+    /// * `dds_config` - The frozen DDS configuration.
+    pub fn new(
+        mut qspi: QspiInterface,
+        io_update_trigger: HighResTimerE,
+        dds_config: DdsConfig,
+    ) -> Self {
+        qspi.start_stream().unwrap();
+        Self {
+            config: dds_config,
+            _qspi: qspi,
+            io_update_trigger,
+        }
+    }
+
+    /// Get a builder for serializing a Pounder DDS profile.
+    pub fn builder(&mut self) -> ProfileBuilder {
+        let builder = self.config.builder();
+        ProfileBuilder {
+            dds_stream: self,
+            serializer: builder,
+        }
+    }
+
+    /// Write a profile to the stream.
+    ///
+    /// # Note:
+    /// If a profile of more than 4 words is provided, it is possible that the QSPI interface will
+    /// stall execution.
+    ///
+    /// # Args
+    /// * `profile` - The serialized DDS profile to write.
+    fn write_profile(&mut self, profile: &[u32]) {
+        // Note(unsafe): We own the QSPI interface, so it is safe to access the registers in a raw
+        // fashion.
+        let regs = unsafe { &*hal::stm32::QUADSPI::ptr() };
+
+        for word in profile.iter() {
+            // Note(unsafe): We are writing to the SPI TX FIFO in a raw manner for performance. This
+            // is safe because we know the data register is a valid address to write to.
+            unsafe {
+                core::ptr::write_volatile(
+                    &regs.dr as *const _ as *mut u32,
+                    *word,
+                );
+            }
+        }
```

Comment:
Before entering this writing loop, should we quickly check the FLEVEL or BUSY bits in the status register to let us know if writing would stall? I'm assuming we don't expect that to ever happen if everything goes as planned.

Reply:
I added a warning if FLEVEL is not zero. I'm unconvinced we should fatally error, but we should inform the user at a minimum.